flocker/testtools/_flaky.py

```diff
@@ -22,10 +37,179 @@ def flaky(jira_key):
 
     # TODO (see FLOC-3281):
     # - allow specifying which exceptions are expected
-    # - retry on expected exceptions
-    #   - parametrize on retry options
     # - provide interesting & parseable logs for flaky tests
 
+    if isinstance(jira_keys, basestring):
+        jira_keys = [jira_keys]
+
+    annotation = _FlakyAnnotation(
+        jira_keys=pset(jira_keys), max_runs=max_runs, min_passes=min_passes)
+
     def wrapper(test_method):
+        setattr(test_method, _FLAKY_ATTRIBUTE, annotation)
         return test_method
+
     return wrapper
+
+
+def _get_flaky_annotation(case):
+    """
+    Get the flaky decoration detail from ``case``.
+
+    :param TestCase case: A test case that might have been decorated with
+        @flaky.
+    :return: ``None`` if not flaky, or a ``pmap`` of the flaky test details.
+    """
+    # XXX: Is there a public way of doing this?
+    method = case._get_test_method()
+    return getattr(method, _FLAKY_ATTRIBUTE, None)
+
+
+class _FlakyAnnotation(PClass):
+
+    max_runs = field(int, mandatory=True,
+                     invariant=lambda x: (x > 0, "must run at least once"))
+    min_passes = field(int, mandatory=True,
+                       invariant=lambda x: (x > 0, "must pass at least once"))
+    jira_keys = pset_field(unicode, optional=False)
+
+    __invariant__ = lambda x: (
+        x.max_runs >= x.min_passes and len(x.jira_keys) > 0,
+        "Can't pass more than we run and must provide a jira key")
+
+    def to_dict(self):
+        return {
+            'max_runs': self.max_runs,
+            'min_passes': self.min_passes,
+            'jira_keys': set(self.jira_keys),
+        }
+
+
+def retry_flaky(run_test_factory=None):
+    """
+    Wrap a ``RunTest`` object so that flaky tests are retried.
+    """
+    if run_test_factory is None:
+        run_test_factory = testtools.RunTest
+
+    return partial(_RetryFlaky, run_test_factory)
+
+
+class _RetryFlaky(testtools.RunTest):
+    """
+    ``RunTest`` implementation that retries tests that fail.
+    """
+    # XXX: This should probably become a part of testtools:
+    # https://bugs.launchpad.net/testtools/+bug/1515933
+
+    def __init__(self, run_test_factory, case, *args, **kwargs):
+        self._run_test_factory = run_test_factory
+        self._case = case
+        self._args = args
+        self._kwargs = kwargs
+
+    def _run_prepared_result(self, result):
+        """
+        Run the test with a result that conforms to testtools' extended
+        ``TestResult`` interface.
+
+        This overrides a method in base ``RunTest`` which is intended to be
+        overwritten.
```

Comment:
testtools actually strives for composition instead of inheritance, but they don't always get it right. I think they'd be open to patches.